### ISSUE TITLE
Paste from Libre Office changes the behaviour of paste filter

### ISSUE THREAD

**Issue by Comandeer** [ckeditor/ckeditor4#3809]
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Open https://codepen.io/Comandeer/pen/MWYZZMN
2. Copy styled `span`.
3. Open local CKEditor sample on `major` branch.
4. Paste copied `span`.

### Expected result

Styles are not preserved (you can see expected result by pasting into the editor inside the pen).

### Actual result

Background and font color are preserved.

## Other details

In Chrome and Safari paste filter is set by default to `semantic-content`, to tidy the polluted HTML produced by these browsers. However the inability to detect the source of paste in Safari and IE resulted in the general filter in PfLO, which changes the behaviour of default paste filter in Blink and WebKit.

* Browser: Chrome, Safari
* CKEditor version: `major`
* Installed CKEditor plugins: `pastefromlibreoffice`

**Comment by f1ames** [comment]
It looks like a serious regression 😭 

From what I understand, we are not able to differentiate content pasted from Libre Office in Safari and IE. The `semantic-content` filter on Chrome/Safari is overwritten by PfLO or there is some logic which just "skips" `semantic-content` filter (what's the order of things here 🤔).

Questions:

1.  Why the PfLO filter also kicks in in Chrome if it was suppose to solve the issue only for Safari/IE?
2.  What can we do about Safari?

:thinking: TBH looks like a hard case since we are not able to distinguish from what source the content was pasted...

**Comment by msamsel** [comment]
> 1. Why the PfLO filter also kicks in in Chrome if it was suppose to solve the issue only for Safari/IE?

The check is not browser specific. We could narrow down the scope of the issue, but still the problematic scenario might appear on Safari and IE:
https://github.com/ckeditor/ckeditor4/blob/01b82d9d86eb1ebabf68934dcf71823f12bf7db8/plugins/pastefromlibreoffice/plugin.js#L44-L45

> 2. What can we do about Safari?

Without meta information in clipboard it won't be possible to distinguish content copied from website from content copied from Libre Office :(
We could apply paste filter in such cases anyway, by making flag `dontFilter` browser related:
https://github.com/ckeditor/ckeditor4/blob/01b82d9d86eb1ebabf68934dcf71823f12bf7db8/plugins/pastefromlibreoffice/plugin.js#L54
However this might result with stripping styles pasted from libre office when those are not defined in pasteFilter.

**Comment by f1ames** [comment]
> The check is not browser specific. We could narrow down the scope of the issue, but still the problematic scenario might appear on Safari and IE

So theoretically we could narrow it down to Safari and IE11 (and older IEs probably) :thinking: For now `canHandle` function always returns true for any `text/html` content which doesn't have `generator` meta tag:

https://github.com/ckeditor/ckeditor4/blob/01b82d9d86eb1ebabf68934dcf71823f12bf7db8/plugins/pastefromlibreoffice/plugin.js#L33-L45

which might result on cases like the above 🤔 So narrowing it down seems like resolving the issue at least on some browsers.

> Without meta information in clipboard it won't be possible to distinguish content copied from website from content copied from Libre Office :(

So it means we couldn't properly handle it on Safari? :fearful: And we have to choose between handling correctly Libre Office content and regression to filtering? 😬

**Comment by msamsel** [comment]
> So it means we couldn't properly handle it on Safari? 😨 And we have to choose between handling correctly Libre Office content and regression to filtering? 😬

Yes. Take a look at clipboard data obtained on Safari for a content with a simple one paragraph:
https://github.com/ckeditor/ckeditor4/blob/01b82d9d86eb1ebabf68934dcf71823f12bf7db8/tests/plugins/pastefromlibreoffice/generated/_fixtures/Simple_text/libreoffice6/safari.html#L1
Content looks like a regular paragraph with a bunch of attributs. None of them seems to be LibreOffice specific, what could allow to distinguish this content from regular copy-paste of an html data.

For Chrome data contains `head` section with `meta` informations, where LibreOffice is clearly defined as content genertor.
https://github.com/ckeditor/ckeditor4/blob/01b82d9d86eb1ebabf68934dcf71823f12bf7db8/tests/plugins/pastefromlibreoffice/generated/_fixtures/Simple_text/libreoffice6/chrome.html#L1-L15

**Comment by f1ames** [comment]
Ok, my proposal is as follows:

1. Narrow the check for Safari and IE only as proposed in https://github.com/ckeditor/ckeditor4/issues/3809#issuecomment-578013826.
1. To not introduce regressions, on Safari we should stick with `semantic-content` filter (so it should as before, filtering spans from Libre Office).
1. Mention in the docs that this is a browser limitation that is hard to workaround and that it works like that by design.

WDYT @Comandeer? @jacekbogdanski any thought on this one?

**Comment by jacekbogdanski** [comment]
Not sure, you probably wrote the same, but if it's only Safari and IE11 (older?) problem - maybe let's just disable PFLO feature for these browsers? We can go with some deeper research later if we can introduce PFLO without regressions for these browsers. We even treat https://github.com/ckeditor/ckeditor4/issues/3646 as a replacement/workaround.

**Comment by Comandeer** [comment]
Both PfLO and PfW explicitly disable paste filter: https://github.com/ckeditor/ckeditor4/blob/01b82d9d86eb1ebabf68934dcf71823f12bf7db8/plugins/pastefromlibreoffice/plugin.js#L54

If we're going to use `semantic-content`, we would probably need to do it manually, right inside PfLO.

What's more, switching on filter produces very poor results.

Libre Office document:
<img width="782" alt="" src="https://user-images.githubusercontent.com/1078728/73251886-14fbf300-41ba-11ea-88e3-d34d0d9ee2bd.png">

Content pasted in Chrome with paste filter switched on:
<img width="967" alt="" src="https://user-images.githubusercontent.com/1078728/73251892-17f6e380-41ba-11ea-8722-c6ec5d948857.png">

> We can go with some deeper research later if we can introduce PFLO without regressions for these browsers

I doubt it. Safari won't expose any metadata (it was already a case in GDocs → https://github.com/ckeditor/ckeditor4/pull/3257#issuecomment-519885641) and LO produces too clean HTML to be able to detect it.

**Comment by f1ames** [comment]
@Comandeer @jacekbogdanski so we basically have a choice between disabling PfLO completely for Safari (as @Comandeer mentioned it's highly improbable something will change in Safari) or go with poor version due to how Safari internally works (mentioning this case in the docs).

Eventually, we could go with #3646 but I doubt we will be able to cover it in a foreseeable future.

**Comment by f1ames** [comment]
After F2F talk with @jacekbogdanski and @Comandeer we have decided that PfLO should be disabled for Safari for now and we will enable it whenever #3646 will be covered.

Leaving this issue open for future reference.
